berkdb/crypto/aes_method.c

```diff
@@ -174,17 +198,40 @@ __aes_derivekeys(dbenv, db_cipher, passwd, plen)
 	if (passwd == NULL) return (EINVAL);
 	AES_CIPHER *aes = (AES_CIPHER *)db_cipher->data;
 
-	SHA_CTX ctx;
-	SHA1_Init(&ctx);
-	SHA1_Update(&ctx, passwd, plen);
-	SHA1_Update(&ctx, (u_int8_t *)DB_ENC_MAGIC, strlen(DB_ENC_MAGIC));
-	SHA1_Update(&ctx, passwd, plen);
-	uint8_t temp[DB_MAC_KEY];
-	SHA1_Final(temp, &ctx);
+	// Create contexts
+	aes->encrypt_ctx = EVP_CIPHER_CTX_new();
+	aes->decrypt_ctx = EVP_CIPHER_CTX_new();
+	
+	if (!aes->encrypt_ctx || !aes->decrypt_ctx)
+	  return (EINVAL);
+
+	// Use Openssl EVP API for SHA1 operations
+	EVP_MD_CTX *md_ctx = EVP_MD_CTX_new();
+	uint8_t key[DB_MAC_KEY];
+	unsigned int key_len;
+	
+	if (!md_ctx)
+	  return (EINVAL);
+	
+	// Initialize EVP digest context with SHA1
+	if (!EVP_DigestInit_ex(md_ctx, EVP_sha1(), NULL) ||
```

Comment:
Noting here that by using SHA-1 we are getting 20 bytes output, which means that using AES-128 and AES_set_encrypt_key and AES_set_decrypt_key will actually truncate the hash output to 16 bytes, resulting in less security than might be possible. I think what I want to do which is a combination of simplicity but still better security is to use SHA-256 instead of SHA-1, with an output of 32 bytes which can be directly input to AES (but moves to AES-256. That might be trickier than I'm hoping...

Reply:
Resolving this because I decided to try and use SHA256 and AES 256 CBC, which worked.